events/service/src/main/java/org/projectnessie/events/service/EventService.java

```diff
@@ -0,0 +1,516 @@
+    LOGGER.debug("Firing {} event: {}", event.getType(), event);
+    forwardToSubscribers(event);
+  }
+
+  private boolean hasContentSubscribers() {
+    return subscribers.hasSubscribersFor(EventType.CONTENT_STORED)
+        || subscribers.hasSubscribersFor(EventType.CONTENT_REMOVED);
+  }
+
+  private boolean hasCommitSubscribers() {
+    return subscribers.hasSubscribersFor(EventType.COMMIT) || hasContentSubscribers();
+  }
+
+  /**
+   * Forwards the event to all subscribers.
+   *
+   * @implNote This implementation is the simplest possible and just invokes all the subscriber
+   *     handlers one by one, synchronously and sequentially. Subclasses may override this method to
+   *     implement a more sophisticated delivery mechanism, e.g. using an asynchronous event bus.
+   */
+  protected void forwardToSubscribers(Event event) {
+    for (EventSubscriberHandler<?> handler : handlers) {
+      handler.deliver(event);
+    }
+  }
+
+  /**
+   * Creates an {@link EventSubscriberHandler} for the given subscriber.
+   *
+   * @implNote This implementation simply creates a new, vanilla, context-less {@link
+   *     EventSubscriberHandler} for the given subscriber. Subclasses may override this method to
+   *     provide custom implementations, e.g. with a tracing context.
+   */
+  protected EventSubscriberHandler<?> createSubscriberHandler(
+      EventSubscriber subscriber, EventSubscription subscription) {
+    return new EventSubscriberHandler<>(subscriber, subscription);
+  }
+
+  /**
+   * A handler for a single subscriber.
+   *
+   * <p>Handlers are responsible for delivering events to the subscriber and retrying failed
+   * delivery attempts.
+   *
+   * <p>Handlers also offer the possibility to pass an arbitrary context object from the caller
+   * thread to the threads doing the retry attempts. This is useful e.g. for propagating traces.
+   *
+   * @param <CTX> The context type for the subscriber.
+   */
+  public class EventSubscriberHandler<CTX> {
+
+    protected final EventSubscriber subscriber;
+    protected final EventSubscription subscription;
+    private final String subscriptionId;
+    private volatile CompletionStage<Void> retryFuture;
+
+    protected EventSubscriberHandler(EventSubscriber subscriber, EventSubscription subscription) {
+      this.subscriber = subscriber;
+      this.subscription = subscription;
+      subscriptionId = subscription.getId().toString();
+    }
+
+    protected void start() {
+      this.subscriber.onSubscribe(subscription);
+    }
+
+    protected void deliver(Event event) {
+      try {
+        MDC.put(SUBSCRIPTION_ID_MDC_KEY, subscriptionId);
+        if (subscriber.accepts(event)) {
+          LOGGER.debug("Received event: {}", event);
+          CTX context = currentContext();
+          Duration initialDelay = config.getRetryConfig().getInitialDelay();
+          if (subscriber.isBlocking()) {
+            deliverAsync(event, 1, initialDelay, null, context);
```

Comment:
But if we implement it later, would it be implemented here, or somewhere else? Trying to figure out if it's best to implement it right now, or decide later if we want this or not.

Reply:
OK I've given this a bit more thought, and I agree to remove the executor completely and the retry mechanism. This will make this service impl completely synchronous by default. But that's enough for this module. In the quarkus-specific module I'd keep the retry logic if you don't mind, but maybe with a default of 0 retries for safety. How does that sound to you?

Reply:
Yea - let's see how that looks in Quarkus